libcxx/include/flat_set

```diff
@@ -0,0 +1,913 @@
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  flat_set(container_type __cont, const key_compare& __comp)
+    : __compare_(__comp), __c_(std::move(__cont)) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(const container_type& __cont, const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a, __cont)) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(container_type&& __cont, const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a, std::move(__cont))) {
+    auto __begin = __c_.begin();
+    auto __end = __c_.end();
+    std::stable_sort(__begin, __end, __comp_ref());
+    __c_.erase(std::__flatset_unique(__begin, __end, __comp_ref()), __end);
+  }
+
+  flat_set(sorted_unique_t, container_type __cont, const key_compare& __comp = key_compare())
+    : __compare_(__comp), __c_(std::move(__cont)) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t, const container_type& __cont, const _Alloc& __a)
+    : __c_(std::__make_obj_using_allocator<container_type>(__a, __cont)) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t, container_type&& __cont, const _Alloc& __a)
+    : __c_(std::__make_obj_using_allocator<container_type>(__a, std::move(__cont))) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t, const container_type& __cont, const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a, __cont)) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t, container_type&& __cont, const key_compare& __comp, const _Alloc& __a)
+    : __compare_(__comp), __c_(std::__make_obj_using_allocator<container_type>(__a, std::move(__cont))) {}
+
+  flat_set(const flat_set& __fs) = default;
+
+  flat_set(flat_set&& __fs)
+    noexcept(is_nothrow_constructible_v<_Compare, _Compare&> && is_nothrow_move_constructible_v<_KeyContainer>)
+    : __compare_(__fs.__compare_), __c_(std::move(__fs.__c_)) {
```

Comment:
Why are you not moving the comparator? You just explained it to me, but please record it on your list of issues.

Reply:
(For the GitHub record: We're copying the comparator because the comparator might be a `std::function<bool(int,int)>`, or a lambda capturing a `std::string`, or anything of that nature. Moving-out-of such a comparator would leave the right-hand `flat_set` in an _invalid_ state — where e.g. `rhs.insert(42)` would crash the program — instead of the _valid but unspecified_ state that is required for conformance.)